oracle/service/runner/runner.go

```diff
@@ -0,0 +1,189 @@
+
+	unsignedVotes := []*oracleproto.Vote{}
+	unsignedVotes = append(unsignedVotes, oracleInfo.UnsignedVoteBuffer.Buffer...)
+
+	oracleInfo.UnsignedVoteBuffer.UpdateMtx.Unlock()
+
+	// sort the votes so that we can rebuild it in a deterministic order, when uncompressing
+	SortOracleVotes(unsignedVotes)
+
+	// batch sign the entire unsignedVoteBuffer and add to gossipBuffer
+	newGossipVote := &oracleproto.GossipedVotes{
+		Validator:       oracleInfo.PubKey.Address(),
+		SignedTimestamp: time.Now().Unix(),
+		Votes:           unsignedVotes,
+	}
+
+	// signing of vote should append the signature field of gossipVote
+	if err := oracleInfo.PrivValidator.SignOracleVote("", newGossipVote); err != nil {
+		log.Errorf("processSignVoteQueue: error signing oracle votes")
+		return
+	}
+
+	// need to mutex lock as it will clash with concurrent gossip
+	preLockTime := time.Now().UnixMilli()
+	oracleInfo.GossipVoteBuffer.UpdateMtx.Lock()
+	address := oracleInfo.PubKey.Address().String()
+	oracleInfo.GossipVoteBuffer.Buffer[address] = newGossipVote
+	oracleInfo.GossipVoteBuffer.UpdateMtx.Unlock()
+	postLockTime := time.Now().UnixMilli()
+	diff := postLockTime - preLockTime
+	if diff > 100 {
+		log.Warnf("WARNING!!! Updating gossip lock took %v milliseconds", diff)
+	}
+}
+
+func PruneVoteBuffers(oracleInfo *types.OracleInfo, consensusState *cs.State) {
+	go func(oracleInfo *types.OracleInfo) {
+		// only keep votes that are less than x blocks old, where x = Config.MaxOracleGossipBlocksDelayed
+		maxOracleGossipBlocksDelayed := oracleInfo.Config.MaxOracleGossipBlocksDelayed
+		// only keep votes that are less than x seconds old, where x = Config.MaxOracleGossipAge
+		maxOracleGossipAge := oracleInfo.Config.MaxOracleGossipAge
+		// run pruner every x milliseconds, where x = Config.PruneInterval
+		pruneInterval := oracleInfo.Config.PruneInterval
+
+		ticker := time.Tick(pruneInterval)
+		for range ticker {
+			lastBlockTime := consensusState.GetState().LastBlockTime.Unix()
+			currTimestampsLen := len(oracleInfo.BlockTimestamps)
+
+			if currTimestampsLen == 0 {
+				oracleInfo.BlockTimestamps = append(oracleInfo.BlockTimestamps, lastBlockTime)
+				continue
+			}
+
+			if oracleInfo.BlockTimestamps[currTimestampsLen-1] != lastBlockTime {
+				oracleInfo.BlockTimestamps = append(oracleInfo.BlockTimestamps, lastBlockTime)
+			}
+
+			// only keep last x number of block timestamps, where x = maxOracleGossipBlocksDelayed
+			if len(oracleInfo.BlockTimestamps) > maxOracleGossipBlocksDelayed {
+				oracleInfo.BlockTimestamps = oracleInfo.BlockTimestamps[1:]
+			}
+
+			latestAllowableTimestamp := time.Now().Unix() - int64(maxOracleGossipAge)
+			// prune votes that are older than the latestAllowableTimestamp, which is the max(earliest block timestamp collected, current time - maxOracleGossipAge)
+			if oracleInfo.BlockTimestamps[0] > latestAllowableTimestamp {
+				latestAllowableTimestamp = oracleInfo.BlockTimestamps[0]
```

Comment:
need to check that buffer is full before replacing

Reply:
resolved in https://github.com/Switcheo/cometbft/pull/1/commits/3babce32ddcfbbf971a0da6717e8f2d1aec5529a